{"compress":true,"commitItems":[["9daaf178-a678-4241-b821-523b41235a2e",1563098007419,"## Java 最常见的 208 道面试题：第七模块答案\n\n[Java团长](javascript:void(0);) _3月24日_\n\n**异常**\n\n**74\\. throw 和 throws 的区别？**\n\nthrows是用来声明一个方法可能抛出的所有异常信息，throws是将异常声明但是不处理，而是将异常往上传，谁调用我就交给谁处理。而throw则是指抛出的一个具体的异常类型。\n\n**75\\. final、finally、finalize 有什么区别？**\n\n*   final可以修饰类、变量、方法，修饰类表示该类不能被继承、修饰方法表示该方法不能被重写、修饰变量表示该变量是一个常量不能被重新赋值。\n\n*   finally一般作用在try-catch代码块中，在处理异常的时候，通常我们将一定要执行的代码方法finally代码块中，表示不管是否出现异常，该代码块都会执行，一般用来存放一些关闭资源的代码。\n\n*   finalize是一个方法，属于Object类的一个方法，而Object类是所有类的父类，该方法一般由垃圾回收器来调用，当我们调用System的gc()方法的时候，由垃圾回收器调用finalize(),回收垃圾。 \n\n**76\\. try-catch-finally 中哪个部分可以省略？**\n\n# 答：catch 可以省略\n\n**原因：**\n\n更为严格的说法其实是：try只适合处理运行时异常，try+catch适合处理运行时异常+普通异常。也就是说，如果你只用try去处理普通异常却不加以catch处理，编译是通不过的，因为编译器硬性规定，普通异常如果选择捕获，则必须用catch显示声明以便进一步处理。而运行时异常在编译时没有如此规定，所以catch可以省略，你加上catch编译器也觉得无可厚非。\n\n理论上，编译器看任何代码都不顺眼，都觉得可能有潜在的问题，所以你即使对所有代码加上try，代码在运行期时也只不过是在正常运行的基础上加一层皮。但是你一旦对一段代码加上try，就等于显示地承诺编译器，对这段代码可能抛出的异常进行捕获而非向上抛出处理。如果是普通异常，编译器要求必须用catch捕获以便进一步处理；如果运行时异常，捕获然后丢弃并且+finally扫尾处理，或者加上catch捕获以便进一步处理。\n\n至于加上finally，则是在不管有没捕获异常，都要进行的“扫尾”处理。\n\n**77\\. try-catch-finally 中，如果 catch 中 return 了，finally 还会执行吗？**\n\n答：会执行，在 return 前执行。\n\n### **代码示例1：**\n\n```\n/*\n```\n\n执行结果：30\n\n### **代码示例2：**\n\n```\npackage com.java_02;\n```\n\n执行结果：40\n\n**78\\. 常见的异常类有哪些？**\n\n*   NullPointerException：当应用程序试图访问空对象时，则抛出该异常。\n\n*   SQLException：提供关于数据库访问错误或其他错误信息的异常。\n\n*   IndexOutOfBoundsException：指示某排序索引（例如对数组、字符串或向量的排序）超出范围时抛出。 \n\n*   NumberFormatException：当应用程序试图将字符串转换成一种数值类型，但该字符串不能转换为适当格式时，抛出该异常。\n\n*   FileNotFoundException：当试图打开指定路径名表示的文件失败时，抛出此异常。\n\n*   IOException：当发生某种I/O异常时，抛出此异常。此类是失败或中断的I/O操作生成的异常的通用类。\n\n*   ClassCastException：当试图将对象强制转换为不是实例的子类时，抛出该异常。\n\n*   ArrayStoreException：试图将错误类型的对象存储到一个对象数组时抛出的异常。\n\n*   IllegalArgumentException：抛出的异常表明向方法传递了一个不合法或不正确的参数。\n\n*   ArithmeticException：当出现异常的运算条件时，抛出此异常。例如，一个整数“除以零”时，抛出此类的一个实例。 \n\n*   NegativeArraySizeException：如果应用程序试图创建大小为负的数组，则抛出该异常。\n\n*   NoSuchMethodException：无法找到某一特定方法时，抛出该异常。\n\n*   SecurityException：由安全管理器抛出的异常，指示存在安全侵犯。\n\n*   UnsupportedOperationException：当不支持请求的操作时，抛出该异常。\n\n*   RuntimeExceptionRuntimeException：是那些可能在Java虚拟机正常运行期间抛出的异常的超类。\n\n**（完）**\n\n**![](data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==)**\n\n**Java团长**\n\n专注于Java干货分享\n\n![](data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==)\n\n扫描上方二维码获取更多Java干货\n\n![](https://mp.weixin.qq.com/mp/qrcode?scene=10000004&size=102&__biz=MzIwMTY0NDU3Nw==&mid=2651938401&idx=1&sn=0f991e835d34488dd24c15dea2d36535&send_time=)\n\n微信扫一扫\n关注该公众号",[[1563097966953,["GJX@GJXAIOU",[[-1,30,"[Java团长](javascript:void(0);) _3月24日_"]],[30,67],[30,30]]],[1563097968595,["GJX@GJXAIOU",[[-1,31,"\n"]],[30,30],[29,29]]],[1563097977192,["GJX@GJXAIOU",[[-1,527,"#"]],[528,528],[527,527]]],[1563097994151,["GJX@GJXAIOU",[[1,1081,"\n"],[1,1083,"\n * java面试题--如果catch里面有return语句，finally里面的代码还会执行吗？\n */\npublic class FinallyDemo2 {\n    public static void main(String[] args) {\n        System.out.println(getInt());\n    }\n\n    public static int getInt() {\n        int a = 10;\n        try {\n            System.out.println(a / 0);\n            a = 20;\n        } catch (ArithmeticException e) {\n            a = 30;\n            return a;\n            /*\n             * return a 在程序执行到这一步的时候，这里不是return a 而是 return 30；这个返回路径就形成了\n             * 但是呢，它发现后面还有finally，所以继续执行finally的内容，a=40\n             * 再次回到以前的路径,继续走return 30，形成返回路径之后，这里的a就不是a变量了，而是常量30\n             */\n        } finally {\n            a = 40;\n        }\n\n//      return a;\n    }\n}"]],[1081,1083],[1770,1770]]],[1563098007920,["GJX@GJXAIOU",[[1,1826,"\n/*\n * java面试题--如果catch里面有return语句，finally里面的代码还会执行吗？\n */\npublic class FinallyDemo2 {\n    public static void main(String[] args) {\n        System.out.println(getInt());\n    }\n\n    public static int getInt() {\n        int a = 10;\n        try {\n            System.out.println(a / 0);\n            a = 20;\n        } catch (ArithmeticException e) {\n            a = 30;\n            return a;\n            /*\n             * return a 在程序执行到这一步的时候，这里不是return a 而是 return 30；这个返回路径就形成了\n             * 但是呢，它发现后面还有finally，所以继续执行finally的内容，a=40\n             * 再次回到以前的路径,继续走return 30，形成返回路径之后，这里的a就不是a变量了，而是常量30\n             */\n        } finally {\n            a = 40;\n            return a; //如果这样，就又重新形成了一条返回路径，由于只能通过1个return返回，所以这里直接返回40\n        }\n\n//      return a;\n    }\n}\n"]],[1805,1825],[2586,2586]]],[1563098019505,["GJX@GJXAIOU",[[-1,3458,"**（完）**"],[-1,3466,"\n**![](data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==)**\n\n**Java团长**\n\n专注于Java干货分享\n\n![](data:image/gif;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVQImWNgYGBgAAAABQABh6FO1AAAAABJRU5ErkJggg==)\n\n扫描上方二维码获取更多Java干货\n\n![](https://mp.weixin.qq.com/mp/qrcode?scene=10000004&size=102&__biz=MzIwMTY0NDU3Nw==&mid=2651938401&idx=1&sn=0f991e835d34488dd24c15dea2d36535&send_time=)\n\n微信扫一扫\n关注该公众号"]],[3458,3933],[3458,3458]]]],null,"GJX@GJXAIOU"],["8998c49d-8e0b-4c74-885d-50632637ed65",1565073238264,"## Java 最常见的 208 道面试题：第七模块答案\n\n\n**异常**\n\n**74\\. throw 和 throws 的区别？**\n\nthrows是用来声明一个方法可能抛出的所有异常信息，throws是将异常声明但是不处理，而是将异常往上传，谁调用我就交给谁处理。而throw则是指抛出的一个具体的异常类型。\n\n**75\\. final、finally、finalize 有什么区别？**\n\n*   final可以修饰类、变量、方法，修饰类表示该类不能被继承、修饰方法表示该方法不能被重写、修饰变量表示该变量是一个常量不能被重新赋值。\n\n*   finally一般作用在try-catch代码块中，在处理异常的时候，通常我们将一定要执行的代码方法finally代码块中，表示不管是否出现异常，该代码块都会执行，一般用来存放一些关闭资源的代码。\n\n*   finalize是一个方法，属于Object类的一个方法，而Object类是所有类的父类，该方法一般由垃圾回收器来调用，当我们调用System的gc()方法的时候，由垃圾回收器调用finalize(),回收垃圾。 \n\n**76\\. try-catch-finally 中哪个部分可以省略？**\n\n 答：catch 可以省略\n\n**原因：**\n\n更为严格的说法其实是：try只适合处理运行时异常，try+catch适合处理运行时异常+普通异常。也就是说，如果你只用try去处理普通异常却不加以catch处理，编译是通不过的，因为编译器硬性规定，普通异常如果选择捕获，则必须用catch显示声明以便进一步处理。而运行时异常在编译时没有如此规定，所以catch可以省略，你加上catch编译器也觉得无可厚非。\n\n理论上，编译器看任何代码都不顺眼，都觉得可能有潜在的问题，所以你即使对所有代码加上try，代码在运行期时也只不过是在正常运行的基础上加一层皮。但是你一旦对一段代码加上try，就等于显示地承诺编译器，对这段代码可能抛出的异常进行捕获而非向上抛出处理。如果是普通异常，编译器要求必须用catch捕获以便进一步处理；如果运行时异常，捕获然后丢弃并且+finally扫尾处理，或者加上catch捕获以便进一步处理。\n\n至于加上finally，则是在不管有没捕获异常，都要进行的“扫尾”处理。\n\n**77\\. try-catch-finally 中，如果 catch 中 return 了，finally 还会执行吗？**\n\n答：会执行，在 return 前执行。\n\n### **代码示例1：**\n\n```\n\n/*\n * java面试题--如果catch里面有return语句，finally里面的代码还会执行吗？\n */\npublic class FinallyDemo2 {\n    public static void main(String[] args) {\n        System.out.println(getInt());\n    }\n\n    public static int getInt() {\n        int a = 10;\n        try {\n            System.out.println(a / 0);\n            a = 20;\n        } catch (ArithmeticException e) {\n            a = 30;\n            return a;\n            /*\n             * return a 在程序执行到这一步的时候，这里不是return a 而是 return 30；这个返回路径就形成了\n             * 但是呢，它发现后面还有finally，所以继续执行finally的内容，a=40\n             * 再次回到以前的路径,继续走return 30，形成返回路径之后，这里的a就不是a变量了，而是常量30\n             */\n        } finally {\n            a = 40;\n        }\n\n//      return a;\n    }\n}\n```\n\n执行结果：30\n\n### **代码示例2：**\n\n```\npackage com.java_02;\n\n/*\n * java面试题--如果catch里面有return语句，finally里面的代码还会执行吗？\n */\npublic class FinallyDemo2 {\n    public static void main(String[] args) {\n        System.out.println(getInt());\n    }\n\n    public static int getInt() {\n        int a = 10;\n        try {\n            System.out.println(a / 0);\n            a = 20;\n        } catch (ArithmeticException e) {\n            a = 30;\n            return a;\n            /*\n             * return a 在程序执行到这一步的时候，这里不是return a 而是 return 30；这个返回路径就形成了\n             * 但是呢，它发现后面还有finally，所以继续执行finally的内容，a=40\n             * 再次回到以前的路径,继续走return 30，形成返回路径之后，这里的a就不是a变量了，而是常量30\n             */\n        } finally {\n            a = 40;\n            return a; //如果这样，就又重新形成了一条返回路径，由于只能通过1个return返回，所以这里直接返回40\n        }\n\n//      return a;\n    }\n}\n```\n\n执行结果：40\n\n**78\\. 常见的异常类有哪些？**\n\n*   NullPointerException：当应用程序试图访问空对象时，则抛出该异常。\n\n*   SQLException：提供关于数据库访问错误或其他错误信息的异常。\n\n*   IndexOutOfBoundsException：指示某排序索引（例如对数组、字符串或向量的排序）超出范围时抛出。 \n\n*   NumberFormatException：当应用程序试图将字符串转换成一种数值类型，但该字符串不能转换为适当格式时，抛出该异常。\n\n*   FileNotFoundException：当试图打开指定路径名表示的文件失败时，抛出此异常。\n\n*   IOException：当发生某种I/O异常时，抛出此异常。此类是失败或中断的I/O操作生成的异常的通用类。\n\n*   ClassCastException：当试图将对象强制转换为不是实例的子类时，抛出该异常。\n\n*   ArrayStoreException：试图将错误类型的对象存储到一个对象数组时抛出的异常。\n\n*   IllegalArgumentException：抛出的异常表明向方法传递了一个不合法或不正确的参数。\n\n*   ArithmeticException：当出现异常的运算条件时，抛出此异常。例如，一个整数“除以零”时，抛出此类的一个实例。 \n\n*   NegativeArraySizeException：如果应用程序试图创建大小为负的数组，则抛出该异常。\n\n*   NoSuchMethodException：无法找到某一特定方法时，抛出该异常。\n\n*   SecurityException：由安全管理器抛出的异常，指示存在安全侵犯。\n\n*   UnsupportedOperationException：当不支持请求的操作时，抛出该异常。\n\n*   RuntimeExceptionRuntimeException：是那些可能在Java虚拟机正常运行期间抛出的异常的超类。\n\n\n",[[1565073183407,["GJX@GJXAIOU",[[1,0,"\n"]],[0,0],[1,1]]],[1565073184249,["GJX@GJXAIOU",[[1,0,"---\nstyle: summer\ntags: [面试]\n---"]],[0,0],[32,32]]],[1565073188728,["GJX@GJXAIOU",[[-1,34,"#"]],[35,35],[34,34]]],[1565073191688,["GJX@GJXAIOU",[[-1,58,"答案"]],[58,60],[58,58]]],[1565073192057,["GJX@GJXAIOU",[[1,58," "],[-1,515," "],[1,516," "],[-1,2803," "],[1,2804," "],[-1,3216," "],[1,3217," "]],[58,58],[59,59]]],[1565073193912,["GJX@GJXAIOU",[[1,59,"yichang"]],[59,59],[66,66]]],[1565073195728,["GJX@GJXAIOU",[[-1,59,"yichang"]],[66,66],[59,59]]],[1565073197474,["GJX@GJXAIOU",[[1,59,"异常"]],[59,59],[61,61]]],[1565073200482,["GJX@GJXAIOU",[[-1,64,"**异常**"]],[64,70],[64,64]]],[1565073201904,["GJX@GJXAIOU",[[-1,66,"**"],[-1,92,"**"]],[80,80],[66,90]]],[1565073203281,["GJX@GJXAIOU",[[-1,181,"**"],[-1,217,"**"]],[196,196],[181,215]]],[1565073204519,["GJX@GJXAIOU",[[-1,507,"**"],[-1,542,"**"]],[523,523],[507,540]]],[1565073207314,["GJX@GJXAIOU",[[-1,990,"**"],[-1,1051,"**"]],[1001,1001],[990,1049]]],[1565073213297,["GJX@GJXAIOU",[[-1,1076,"**"],[-1,1072,"### **"]],[1080,1080],[1072,1072]]],[1565073218239,["GJX@GJXAIOU",[[-1,1792,"**"],[-1,1788,"### **"]],[1798,1798],[1788,1788]]],[1565073222466,["GJX@GJXAIOU",[[-1,2596,"**"],[-1,2613,"**"]],[2607,2607],[2596,2611]]],[1565073231583,["GJX@GJXAIOU",[[-1,183,"\\"],[1,184," "]],[183,184],[183,184]]],[1565073232183,["GJX@GJXAIOU",[[-1,509,"\\"],[1,510," "]],[509,510],[509,510]]],[1565073232593,["GJX@GJXAIOU",[[-1,992,"\\"],[1,993," "]],[992,993],[992,993]]],[1565073233079,["GJX@GJXAIOU",[[-1,2598,"\\"],[1,2599," "]],[2598,2599],[2598,2599]]],[1565073233920,["GJX@GJXAIOU",[[-1,68,"\\"],[1,69," "]],[68,69],[68,69]]],[1565073236680,["GJX@GJXAIOU",[[1,66,"## "]],[71,71],[93,93]]],[1565073298285,[null,[[-1,1075,"### "],[-1,1085,"**"],[-1,1795,"### **"],[1,1801,"\n\n"],[-1,1807,"**"],[1,2599,"`\n"],[-1,2601," "],[1,2602,"行"],[-1,2609,"**"],[-1,2613,"\\"],[1,2614," . "],[-1,2624,"**"]],[1075,1075],[2624,2624]]],[1565073298285,[null,[[1,1075,"### "],[1,1081,"**"],[1,1789,"### **"],[-1,1789,"\n\n"],[1,1797,"**"],[-1,2587,"`\n"],[1,2591," "],[-1,2591,"行"],[1,2599,"**"],[1,2601,"\\"],[-1,2601," . "],[1,2614,"**"]],[2624,2624],[1075,1075]]],[1565073239104,["GJX@GJXAIOU",[[1,66,"\n"]],[62,62],[63,63]]],[1565073241538,["GJX@GJXAIOU",[[1,63,"@toc"]],[63,63],[67,67]]],[1565073244826,["GJX@GJXAIOU",[[-1,70,"\n"]],[70,70],[69,69]]],[1565073247266,["GJX@GJXAIOU",[[1,188,"## "]],[195,195],[225,225]]],[1565073250138,["GJX@GJXAIOU",[[1,533,"2"]],[533,533],[534,534]]],[1565073251424,["GJX@GJXAIOU",[[-1,533,"2"]],[534,534],[533,533]]],[1565073251760,["GJX@GJXAIOU",[[1,517,"## "]],[533,533],[553,553]]],[1565073254008,["GJX@GJXAIOU",[[1,1003,"## "]],[1021,1021],[1065,1065]]],[1565073258321,["GJX@GJXAIOU",[[1,2612,"## "]],[2622,2622],[2630,2630]]],[1565073267689,["GJX@GJXAIOU",[[1,1804,"- "]],[1804,1804],[1806,1806]]],[1565073270303,["GJX@GJXAIOU",[[-1,1794,"\n"]],[1794,1794],[1793,1793]]],[1565073276463,["GJX@GJXAIOU",[[1,1538,"\n             "]],[1538,1538],[1552,1552]]],[1565073280536,["GJX@GJXAIOU",[[1,1578,"\n             "]],[1578,1578],[1592,1592]]],[1565073282696,["GJX@GJXAIOU",[[1,1637,"\n             "]],[1637,1637],[1651,1651]]],[1565073284154,["GJX@GJXAIOU",[[1,1711,"\n             "]],[1711,1711],[1725,1725]]],[1565073285905,["GJX@GJXAIOU",[[1,1746,"\n             "]],[1746,1746],[1760,1760]]],[1565073358285,[null,[[-1,1088,"### "],[-1,1098,"**"],[1,1538,"一步的时候，"],[-1,1552,"一步的时候，"],[-1,1578,"\n "],[1,1580,"rn 30；\n"],[-1,1592,"rn"],[-1,1595,"30；"],[1,1637,"nally，"],[-1,1651,"nally，"],[1,1711,"rn 30，"],[-1,1725,"rn 30，"],[1,1746,"是a变量了，"],[-1,1760,"是a变量了，"],[-1,1864,""],[1,1864,"}"],[-1,1869,"\n"],[-1,1873,"- "],[-1,1879,"### **"],[1,1885,"\n\n- "],[-1,1891,"**"],[-1,2683,"## \n执 "],[1,2689,"`\n\n执行"],[-1,2696,"**78\\"],[1,2701,"## 78 . "],[-1,2711,"**"]],[1088,1088],[2711,2711]]],[1565073358285,[null,[[1,1088,"### "],[1,1094,"**"],[-1,1532,"一步的时候，"],[1,1552,"一步的时候，"],[1,1572,"\n "],[-1,1572,"rn 30；\n"],[1,1591,"rn"],[1,1592,"30；"],[-1,1631,"nally，"],[1,1651,"nally，"],[-1,1705,"rn 30，"],[1,1725,"rn 30，"],[-1,1740,"是a变量了，"],[1,1760,"是a变量了，"],[1,1858,""],[-1,1858,"}"],[1,1864,"\n"],[1,1867,"- "],[1,1871,"### **"],[-1,1871,"\n\n- "],[1,1881,"**"],[1,2671,"## \n执 "],[-1,2671,"`\n\n执行"],[1,2683,"**78\\"],[-1,2683,"## 78 . "],[1,2701,"**"]],[2711,2711],[1088,1088]]],[1565073305041,["GJX@GJXAIOU",[[-1,1539,"             "]],[1552,1552],[1539,1539]]],[1565073305536,["GJX@GJXAIOU",[[-1,1538,"\n"]],[1539,1539],[1538,1538]]],[1565073316121,["GJX@GJXAIOU",[[-1,1624,"             "]],[1624,1637],[1624,1624]]],[1565073316546,["GJX@GJXAIOU",[[-1,1623,"\n"]],[1624,1624],[1623,1623]]],[1565073319577,["GJX@GJXAIOU",[[-1,1684,"             "]],[1684,1697],[1684,1684]]],[1565073319976,["GJX@GJXAIOU",[[-1,1683,"\n"]],[1684,1684],[1683,1683]]]],null,"GJX@GJXAIOU"]]}